Adding legacy USB driver

UtsavAgarwalADI opened the pull request:
## PR Description

Adding legacy musb driver for sc57x/58x and enabling DMA where applicable

## PR Type
- [x] Bug fix (a change that fixes an issue)
- [x] New feature (a change that adds new functionality)
- [ ] Breaking change (a change that affects other repos or cause CIs to fail)

## PR Checklist
- [x] I have conducted a self-review of my own code changes
- [x] I have tested the changes on the relevant hardware
- [ ] I have updated the documentation outside this repo accordingly (if there is the case)

pamolloy commented on Documentation/devicetree/bindings/usb/adi,musb.yaml:
```diff
@@ -0,0 +1,85 @@
+
+maintainers:
+  - Nathan Barrett-Morrison <nathan.morrison@timesys.com>
+  - Greg Malysa <greg.malysa@timesys.com>
```
Do they still want to maintain these if they ever make it upstream? Either way, I assume these e-mail addresses aren't valid anymore?

artursartamonovsadi replied:
probably should be just adsp-linux@analog.com

pamolloy commented on drivers/usb/musb/adi.h:
```diff
@@ -0,0 +1,23 @@
+ *
+ * Contact: Nathan Barrett-Morrison <nathan.morrison@timesys.com>
+ * Contact: Greg Malysa <greg.malysa@timesys.com>
+ *
```
```suggestion
```

pamolloy commented on drivers/usb/musb/adi.c:
```diff
@@ -0,0 +1,424 @@
+module_platform_driver(adi_musb_driver);
+
+MODULE_DESCRIPTION("ADI MUSB Glue Layer");
+MODULE_AUTHOR("Hao Liang <hliang1025@gmail.com>");
```
Curious, I see one mainline contribution from Hao Liang:

```
1569039db006 bfin uart: it will hang when read current y count if not disable dma irq
```

I wonder why he's listed as the author.

pamolloy commented on arch/arm/configs/sc589-mini_defconfig:
```diff
@@ -120,4 +120,6 @@ CONFIG_CRC_CCITT=y
 CONFIG_DEBUG_KERNEL=y
 CONFIG_DEBUG_FS=y
+CONFIG_USB_MUSB_ADI=y
+CONFIG_USB_INVENTRA_DMA=y
```
Looks like there is a conflict, but also it is good practice to select through `menuconfig` and `savedefconfig` or to run `make sc5890-mini_defconfig; make savedefconfig` before committing.

Ideally the commit message would be declarative and have a prefix matching mainline:

```
ARM: configs: sc589-mini: Enable MUSB DMA
```

`ARM: defconfig:` would refer to multiple different defconfigs.

nunojsa commented on drivers/usb/musb/adi.c:
```diff
@@ -0,0 +1,424 @@
+		   .of_match_table = adi_musb_match,
+		   },
+};
+
```
nit: no need for blank line

nunojsa commented on drivers/usb/musb/adi.h:
```diff
@@ -0,0 +1,23 @@
+#define __ADI_H__
+
+static int get_int_prop(struct device_node *node, const char *s);
+
```
static?

nunojsa commented on drivers/usb/musb/adi.c:
```diff
@@ -0,0 +1,424 @@
+
+	platform_device_unregister(glue->musb);
+	usb_phy_generic_unregister(glue->phy);
+	kfree(glue);
```
fairly sure the above adi_musb_remove() is not needed

nunojsa commented on drivers/usb/musb/adi.c:
```diff
@@ -0,0 +1,424 @@
+	kfree(glue);
+
+err0:
+	dev_err(&pdev->dev, "ADI MUSB device FAILED\n");
```
use device managed APIs and you should be able to remove all the above labels.

nunojsa commented on drivers/usb/musb/adi.c:
```diff
@@ -0,0 +1,424 @@
+	{},
+};
+
+MODULE_DEVICE_TABLE(of, adi_musb_match);
```
This should be closer to the driver initialization

nunojsa commented on drivers/usb/musb/adi.c:
```diff
@@ -0,0 +1,424 @@
+	.set_mode = adi_musb_set_mode,
+};
+
+static u64 musb_dmamask = DMA_BIT_MASK(32);
```
I believe the above is already the default

nunojsa commented on drivers/usb/musb/adi.c:
```diff
@@ -0,0 +1,424 @@
+		break;
+	default:
+		dev_err(&parent->dev, "Trying to set unsupported mode %u\n",
+				musb_mode);
```
error should be returned

nunojsa commented on drivers/usb/musb/adi.c:
```diff
@@ -0,0 +1,424 @@
+
+	ret = platform_get_irq(pdev, 1);
+	if (!ret)
+		dev_err(&pdev->dev, "failed to get dma irq.\n");
```
no return?

nunojsa commented on drivers/usb/musb/adi.c:
```diff
@@ -0,0 +1,424 @@
+
+	class = get_int_prop(pdev->dev.of_node, "mode");
+	if (class == MUSB_OTG) {
+#if defined(CONFIG_USB_MUSB_HOST)
```
IS_ENABLED()

nunojsa commented on drivers/usb/musb/adi.c:
```diff
@@ -0,0 +1,424 @@
+	config = devm_kzalloc(&pdev->dev, sizeof(*config), GFP_KERNEL);
+	if (!config) {
+		ret = -ENOMEM;
+		goto err2;
```
return dev_err_probe(). Ditto for all other places

nunojsa commented on drivers/usb/musb/adi.c:
```diff
@@ -0,0 +1,424 @@
+
+	musb->dev.parent	= &pdev->dev;
+	musb->dev.dma_mask	= &musb_dmamask;
+	musb->dev.coherent_dma_mask = musb_dmamask;
```
nit: use spaces

nunojsa commented on drivers/usb/musb/adi.c:
```diff
@@ -0,0 +1,424 @@
+	}
+
+	/*alloc musb platform_device and register */
+	musb = platform_device_alloc("musb-hdrc", PLATFORM_DEVID_AUTO);
```
no new line

nunojsa commented on drivers/usb/musb/adi.c:
```diff
@@ -0,0 +1,424 @@
+	if (!match) {
+		dev_err(&pdev->dev, "failed to matching of_match node\n");
+		return -EINVAL;
+	}
```
Why having the above?

nunojsa commented on drivers/usb/musb/adi.c:
```diff
@@ -0,0 +1,424 @@
+		return 0;
+
+	return val;
+}
```
The above helper does not bring much

nunojsa reviewed (commented):
Some initial comments. The commit message should be improved a lot... That you're adding some glue code we can see :). You need to explain why we need it for our platforms and some explanation on why all those platforms devices are needed would also be great.

nunojsa reviewed (commented):
Some initial comments. The commit message should be improved a lot... That you're adding some glue code we can see :). You need to explain why we need it for our platforms and some explanation on why all those platforms devices are needed would also be great.

UtsavAgarwalADI commented:
**Changelog:** 

**Driver:**
- Replaced maintainer info
- Removed `get_int_prop()` due to lack of any meaningful contribution
- Removed unnecessary static function declaration within header
- Removed explicit dma mask value assignment since we already use the default
- `adi_musb_set_mode()` now returns errors upon failure
- Removed node matching from probe which does not contribute to device functionality
- Replaced `#defined()` with `#IF_ENABLED()`
- Updated error handling within probe to utilize `dev_err_probe()`
- Updated error handling within probe to unregister phy upon failure as well
- Updated commit description to explain why this driver is required for the kernel
- Updated maintainer info

**Documentation:**
- Updated maintainer info

**Configs:**
- Updated SC589-mini defconfig with options generated from `make savedefconfig`
- Updated commit subject from defconfig to configs to indicate single platform config change

UtsavAgarwalADI commented:
**Changelog**:

Adding signature to last commit
